File: 백준/Gold/13549. 숨바꼭질 3/숨바꼭질 3.py
import sys
import math
from collections import deque

input = sys.stdin.readline

N, K = map(int, input().split())

if K <= N:
    print(N - K)
    exit()

MAX_LENGTH = 100001

arr = [math.inf] * MAX_LENGTH
visited = [False] * MAX_LENGTH

arr[N] = 0
visited[N] = True

que = deque([N])

while len(que) > 0:
    cur = que.popleft()

    if cur * 2 < MAX_LENGTH and not visited[cur * 2]:
        que.appendleft(cur * 2)
        arr[cur * 2] = arr[cur]
        visited[cur * 2] = True

    if cur - 1 > 0 and not visited[cur - 1]:
        que.append(cur - 1)
        arr[cur - 1] = arr[cur] + 1
        visited[cur - 1] = True

    if cur + 1 < MAX_LENGTH and not visited[cur + 1]:
        que.append(cur + 1)
        arr[cur + 1] = arr[cur] + 1
        visited[cur + 1] = True

print(arr[K])

File: 백준/Gold/13549. 숨바꼭질 3/숨바꼭질 3.js
let [N, K] = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split(" ")
  .map(Number);

const MAX_LENGTH = 100001;

let arr = Array(MAX_LENGTH).fill(Infinity);
let visited = Array(MAX_LENGTH).fill(false);

arr[N] = 0;
visited[N] = true;

let que = [N];

while (que.length > 0) {
  let cur = que.shift();

  if (cur * 2 < MAX_LENGTH && !visited[cur * 2]) {
    que.unshift(cur * 2);
    arr[cur * 2] = arr[cur];
    visited[cur * 2] = true;
  }

  if (cur - 1 >= 0 && !visited[cur - 1]) {
    que.push(cur - 1);
    arr[cur - 1] = arr[cur] + 1;
    visited[cur - 1] = true;
  }

  if (cur + 1 < MAX_LENGTH && !visited[cur + 1]) {
    que.push(cur + 1);
    arr[cur + 1] = arr[cur] + 1;
    visited[cur + 1] = true;
  }
}

console.log(arr[K]);
